hub/core/transform/transform.py

```diff
@@ -0,0 +1,166 @@
+        InvalidTransformException: If the output of any step in a transformation isn't dictionary or a list/tuple of dictionaries.
+    """
+    ds_out.flush()
+
+    if not hasattr(data_in, "__getitem__"):
+        raise InvalidInputDataException("__getitem__")
+    if not hasattr(data_in, "__len__"):
+        raise InvalidInputDataException("__len__")
+
+    # TODO: this check doesn't work currently. Will work once AL-1092 is merged and can be uncommented.
+    # for tensor in tensors:
+    #     assert len(ds_out[tensor]) == len(ds_out)
+
+    workers = max(workers, 1)
+
+    if scheduler == "threaded":
+        compute: ComputeProvider = ThreadProvider(workers)
+    elif scheduler == "processed":
+        compute = ProcessProvider(workers)
+    else:
+        raise UnsupportedSchedulerException(scheduler)
+
+    base_storage = remove_all_cache(ds_out.storage)
+    tensors = set(ds_out.meta.tensors)
+
+    pipeline_kwargs = pipeline_kwargs or []
+    pipeline_kwargs = pipeline_kwargs[0 : len(pipeline)]
+    pipeline_kwargs += [{}] * (len(pipeline) - len(pipeline_kwargs))
+
+    store(data_in, base_storage, tensors, compute, workers, pipeline, pipeline_kwargs)
+    ds_out.clear_cache()
+    ds_out._load_meta()
+
+
+def store_shard(transform_input: Tuple):
+    """Takes a shard of the original data and iterates through it, producing chunks."""
+    data_shard, size, storage, tensors, pipeline, pipeline_kwargs = transform_input
+
+    # storing the metas in memory to merge later
+    all_index_meta = {key: IndexMeta.create(key, MemoryProvider()) for key in tensors}
+    all_tensor_meta = {key: TensorMeta.create(key, MemoryProvider()) for key in tensors}
+
+    # separate cache for each tensor to prevent frequent flushing, 32 MB ensures only full chunks are written.
+    storage_map = {key: LRUCache(MemoryProvider(), storage, 32 * MB) for key in tensors}
+
+    # will be simply range(len(data_shard)) after AL 1092
+    for i in range(min(len(data_shard), size)):
+        sample = data_shard[i]
+        if isinstance(sample, Dataset):
+            sample_dict = {}
+            for k in sample.tensors:
+                sample_dict[k] = sample[k].numpy()
+            sample = sample_dict
+
+        # always a list of dicts
+        results = transform_sample(sample, pipeline, pipeline_kwargs)
+        for result in results:
+            if set(result.keys()) != tensors:
+                raise TensorMismatchException(list(tensors), list(result.keys()))
+            for key, value in result.items():
+                append_tensor(
+                    value,
+                    key,
+                    storage_map[key],
+                    index_meta=all_index_meta[key],
+                    tensor_meta=all_tensor_meta[key],
+                )
+
+    for tensor in tensors:
+        storage_map[tensor].flush()
+
+    return all_index_meta, all_tensor_meta
+
+
+def store(data_in, storage, tensors, compute, workers, pipeline, pipeline_kwargs):
```

Comment:
docstring pls

Reply:
if you have `TransformInput` namedtuple, this entire parameter set can be reduced to just `transform_input` with a static type `TransformInput`